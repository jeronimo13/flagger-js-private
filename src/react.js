import PropTypes from 'prop-types'
import React from 'react'

import Airship from './index'
export {Airship}

export class FlagProvider extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      loading: !Airship.environment
    }

    if (Airship.environment) {
      this.configure()
    }
  }

  async configure() {
    if (!Airship.environment) {
      await Airship.configure({
        envKey: this.props.envKey,
        flagConfig: this.props.flagConfig,
        subscribeToUpdates: this.props.subscribeToUpdates
      })
    }
    if (this.props.entity) {
      Airship.identify(this.props.entity)
    }
  }

  async componentDidMount() {
    if (Airship.environment) {
      return
    }

    await this.configure()
    this.setState({loading: false})
  }

  render() {
    return this.state.loading ? this.props.loadingView : this.props.children
  }
}

FlagProvider.propTypes = {
  envKey: PropTypes.string,
  entity: PropTypes.object,
  flagConfig: PropTypes.object,
  loadingView: PropTypes.node,
  subscribeToUpdates: PropTypes.bool,
  children: PropTypes.node
}

FlagProvider.defaultProps = {
  children: null,
  loadingView: null,
  subscribeToUpdates: false
}

export function withFlag(WrappedComponent, ...flagNames) {
  return class FlaggedComponent extends React.Component {
    componentDidMount() {
      Airship.addGatingInfoListener(this.handleChange)
    }

    componentWillUnmount() {
      Airship.removeGatingInfoListener(this.handleChange)
    }

    handleChange = () => {
      this.forceUpdate()
    }

    render() {
      if (flagNames.length === 0) {
        // eslint-disable-next-line no-console
        console.warn('withFlag did not receive a valid flag name')
        return <WrappedComponent flags={{}} {...this.props} />
      }

      const flags = flagNames.reduce((flags, flagName) => {
        const flag = Airship.flag(flagName)
        const enabled = flag.isEnabled(this.props.entity)
        const eligible = flag.isEligible(this.props.entity)
        const treatment = flag.getTreatment(this.props.entity)
        const payload = flag.getPayload(this.props.entity)

        return {
          ...flags,
          [flagName]: {
            enabled,
            eligible,
            treatment,
            payload
          }
        }
      }, {})

      return <WrappedComponent flags={flags} {...this.props} />
    }
  }
}

export class FlagSwitch extends React.Component {
  render() {
    return React.Children.map(this.props.children, child => {
      if (child.type.prototype instanceof Flag || child.type === Flag) {
        const props = {}
        if (!child.props.flag) {
          props.flag = this.props.flag
        }
        if (!child.props.entity && this.props.entity) {
          props.entity = this.props.entity
        }
        return React.cloneElement(child, props, child.props.children)
      } else {
        return child
      }
    })
  }
}

FlagSwitch.propTypes = {
  flag: PropTypes.string.isRequired,
  children: PropTypes.node,
  entity: PropTypes.object
}

export class Flag extends React.Component {
  componentDidMount() {
    Airship.addGatingInfoListener(this.handleChange)
  }

  componentWillUnmount() {
    Airship.removeGatingInfoListener(this.handleChange)
  }

  handleChange = () => {
    this.forceUpdate()
  }

  render() {
    if (!this.props.flag) {
      // eslint-disable-next-line no-console
      console.warn('<Flag> component missing flag name')
      return null
    }

    const treatment = Airship.flag(this.props.flag).getTreatment(
      this.props.entity
    )
    if (treatment !== this.props.case) {
      return null
    }

    return this.props.children
  }
}

Flag.propTypes = {
  flag: PropTypes.string,
  case: PropTypes.string.isRequired,
  children: PropTypes.node,
  entity: PropTypes.object
}

Flag.defaultProps = {
  children: null
}
