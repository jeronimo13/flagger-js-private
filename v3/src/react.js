import React from 'react'

export class FlagProvider extends React.Component {
  render() {
    return <div>Test React Component</div>
  }
}
