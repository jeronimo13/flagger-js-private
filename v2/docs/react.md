# Flagger for React

## Introduction

Flagger is an open-source library for feature flagging. Feature flagging allows teams to deliver features quickly and safely by giving them fine-grained control over when and which users are able to access new features.

## Concepts

Before using this library, it will help to understand some of the [concepts documented](concepts.md).

## Installation

In a browser:

```html
<script src="https://d2or86eailfls8.cloudfront.net/2/flagger-react.js"></script>
```

Using npm:

```javascript
npm install --save flagger
```

In Node.js:

```javascript
const {FlagProvider, FlagSwitch, Flag, withFlag} = require('flagger/react')
```

In ES2015 modules:

```javascript
import {FlagProvider, FlagSwitch, Flag, withFlag} from 'flagger/react'
```

## API

### `<FlagProvider>`

Configures the rules used for feature flagging.

```jsx
<FlagProvider
  envKey={optionalString}
  flagConfig={optionalObject}
  entity={optionalObject}
  loadingView={optionalNode}
  subscribeToUpdates={optionalBoolean}>
  <App />
</FlagProvider>
```

**This should be top-level singleton in your application. Do not create more than one FlagProvider.**

**envKey: string**

The environment key used to connect to Airship (paid)

**flagConfig: object**

The locally provided flag rules to use with Flagger (free). Structure of the object documented [here](flagconfig.md).

**entity: object**

The default entity to check the flag treatment for.

**loadingView: node**

What to show while fetching flag rules from Airship.

**subscribeToUpdates: bool**

Whether to subscribe to updates to flag rules from Airship and automatically re-render when the flag rules change. Defaults to `false`.

### `<FlagSwitch>`

Fills in missing `flag` and optionally `entity` properties for `<Flag>` elements, **which must be direct children**.

```jsx
<FlagSwitch flag={requiredString} entity={optionalEntity}>
  <Flag ... />
</FlagSwitch>
```

**flag: string**

The name of the flag.

**entity: object**

The entity to check the flag treatment for.

### `<Flag>`

Renders children based on whether the `case` prop matches the flag treatment for the `entity`.

```jsx
<Flag flag={optionalString} case={requiredString} entity={optionalObject}>
  Content to render when case matches flag treatment
</Flag>
```

- **`flag` prop is only optional when `<Flag>` is a direct child of a `<FlagSwitch>` element**.
- **`entity` is not optional if it is not inherited from `<FlagProvider>` or `<FlagSwitch>`**.
- `<Flag>` does not have to be a child of `<FlagSwitch>` and can be used standalone.

**flag: string**

The name of the flag.

**case: string**

The name of the flag treatment.

**entity: object**

The entity to check the flag treatment for.

### Examples

```jsx
<FlagProvider envKey="onz2150xjon6pkjr">
  <FlagSwitch flag="bitcoin-pay">
    <Flag entity={{id: 1}} case="on">
      <BitcoinPaymentButton />
    </Flag>
  </FlagSwitch>
</FlagProvider>
```

```jsx
<FlagProvider envKey="onz2150xjon6pkjr">
  <FlagSwitch flag="color-theme" entity={{id: 1}}>
    <Flag case="blue">
      <Background color="blue" />
    </Flag>
    <Flag case="lavender">
      <Background color="lavender" />
    </Flag>
    <Flag case="off">
      <Background color="default" />
    </Flag>
  </FlagSwitch>
</FlagProvider>
```

## Server side rendering

Since `FlagProvider` is an asynchronous component that requires time to configure. You may notice that it always renders the `loadingView` on server side rendering. To get around this you can configure Flagger ahead of time, so that `FlagProvider` becomes a synchronous component.

To do so, import Flagger from Flagger React and call configure on it:

```js
import {Flagger} from 'flagger/react'

await Flagger.configure({ ... })
```

You can read more documentation about `Flagger.configure` [here](api-reference.md#flaggerconfigureoptions).

You will need to configure ahead of time both on the server side and client side, so that the first initial render on the client side does not show `loadingView`.

On the server, we recommend passing in the option `{ subscribeToUpdates: true }`, while on the client side we recommend not doing so, so that the client side does not automatically re-render on hosted rule changes, which may be a jarring user experience.
