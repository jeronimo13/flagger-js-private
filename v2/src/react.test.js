import nock from 'nock'
import React from 'react'
import renderer from 'react-test-renderer'

import Airship from './index'
import {withFlag, FlagSwitch, Flag} from './react'

beforeAll(async () => {
  nock('https://api.airshiphq.com', {encodedQueryParams: true})
    .get('/v2/gating-info/onz2150xjon6pkjr')
    .query({casing: 'camel'})
    .reply(200, {
      flags: [
        {
          splits: [{treatmentId: '36', percentage: 1}],
          populations: [
            {
              rules: [],
              universes: [],
              percentage: 0,
              entityType: 'User',
              hashKey: '1'
            }
          ],
          overrides: [{treatmentId: '36', entityId: '1', entityType: 'User'}],
          treatments: [
            {
              treatmentId: '35',
              codename: 'off',
              isControl: false,
              isOffTreatment: true
            },
            {
              treatmentId: '36',
              codename: 'on',
              isControl: false,
              isOffTreatment: false,
              payload: {
                a: 'b'
              }
            }
          ],
          flagType: 'basic',
          hashKey: '16',
          isPaused: false,
          isWebAccessible: true,
          codename: 'bitcoin-pay',
          flagStatus: 'operational'
        },
        {
          splits: [
            {treatmentId: '40', percentage: 0},
            {treatmentId: '38', percentage: 1},
            {treatmentId: '39', percentage: 0}
          ],
          populations: [
            {
              rules: [],
              universes: [],
              percentage: 0,
              entityType: 'User',
              hashKey: '2'
            }
          ],
          overrides: [
            {treatmentId: '38', entityId: '1', entityType: 'User'},
            {treatmentId: '39', entityId: '2', entityType: 'User'},
            {treatmentId: '40', entityId: '3', entityType: 'User'}
          ],
          treatments: [
            {
              treatmentId: '37',
              codename: 'off',
              isControl: false,
              isOffTreatment: true
            },
            {
              treatmentId: '38',
              codename: 'default',
              isControl: false,
              isOffTreatment: false
            },
            {
              treatmentId: '39',
              codename: 'variation-1',
              isControl: false,
              isOffTreatment: false
            },
            {
              treatmentId: '40',
              codename: 'variation-2',
              isControl: false,
              isOffTreatment: false
            }
          ],
          flagType: 'basic',
          hashKey: '17',
          isPaused: false,
          isWebAccessible: true,
          codename: 'treatment',
          flagStatus: 'operational'
        }
      ],
      env: {hashKey: '1', envKey: 'onz2150xjon6pkjr'},
      sdkInfo: {SDK_INGESTION_INTERVAL: 60, SDK_INGESTION_MAX_ITEMS: 500}
    })
  nock.disableNetConnect()

  await Airship.configure({envKey: 'onz2150xjon6pkjr'})
  Airship.identify({id: 1})
})

afterAll(() => {
  nock.cleanAll()
  nock.enableNetConnect()
})

test('withFlag adds flags.<name>.enabled property', () => {
  function TestComponent(props) {
    return <div>{props.flags['bitcoin-pay'].enabled.toString()}</div>
  }

  const WrappedComponent = withFlag(TestComponent, 'bitcoin-pay')

  const component = renderer.create(<WrappedComponent />)

  expect(component.toJSON()).toMatchSnapshot()
})

test('withFlag adds flags.<name>.eligible property', () => {
  function TestComponent(props) {
    return <div>{props.flags['bitcoin-pay'].eligible.toString()}</div>
  }

  const WrappedComponent = withFlag(TestComponent, 'bitcoin-pay')

  const component = renderer.create(<WrappedComponent />)

  expect(component.toJSON()).toMatchSnapshot()
})

test('withFlag adds flags.<name>.treatment property', () => {
  function TestComponent(props) {
    return <div>{props.flags['bitcoin-pay'].treatment.toString()}</div>
  }
  const WrappedComponent = withFlag(TestComponent, 'bitcoin-pay')

  const component = renderer.create(<WrappedComponent />)

  expect(component.toJSON()).toMatchSnapshot()
})

test('withFlag adds flags.<name>.payload property', () => {
  function TestComponent(props) {
    return <div>{JSON.stringify(props.flags['bitcoin-pay'].payload)}</div>
  }
  const WrappedComponent = withFlag(TestComponent, 'bitcoin-pay')

  const component = renderer.create(<WrappedComponent />)

  expect(component.toJSON()).toMatchSnapshot()
})

test('withFlag wrapped components accept entity prop', () => {
  function TestComponent(props) {
    return <div>{props.flags['bitcoin-pay'].enabled.toString()}</div>
  }
  const WrappedComponent = withFlag(TestComponent, 'bitcoin-pay')

  const component = renderer.create(<WrappedComponent entity={{id: 2}} />)

  expect(component.toJSON()).toMatchSnapshot()
})

test('withFlag wrapped components handles invalid flag name', () => {
  function TestComponent(props) {
    return <div>{props.flags['unreal-flag'].enabled.toString()}</div>
  }
  const WrappedComponent = withFlag(TestComponent, 'unreal-flag')

  const component = renderer.create(<WrappedComponent entity={{id: 2}} />)

  expect(component.toJSON()).toMatchSnapshot()
})

test('FlagSwitch accepts entity prop', () => {
  const component = renderer.create(
    <FlagSwitch flag="bitcoin-pay" entity={{id: 2}}>
      <Flag case="on">On!</Flag>
      <Flag case="off">Off!</Flag>
    </FlagSwitch>
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('Flag can inherit flag prop from FlagSwitch', () => {
  const component = renderer.create(
    <FlagSwitch flag="bitcoin-pay">
      <Flag case="on">On!</Flag>
      <Flag case="off">Off!</Flag>
    </FlagSwitch>
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('Flag can override flag prop from FlagSwitch', () => {
  const component = renderer.create(
    <FlagSwitch flag="bitcoin-pay">
      <Flag flag="treatment" case="default">
        Default
      </Flag>
      <Flag flag="treatment" case="variation-1">
        Variation 1
      </Flag>
      <Flag flag="treatment" case="variation-2">
        Variation 1
      </Flag>
      <Flag flag="treatment" case="off">
        Off
      </Flag>
    </FlagSwitch>
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('Flag accepts entity prop', () => {
  const component = renderer.create(
    <>
      <Flag flag="treatment" entity={{id: 2}} case="default">
        Default
      </Flag>
      <Flag flag="treatment" entity={{id: 2}} case="variation-1">
        Variation 1
      </Flag>
      <Flag flag="treatment" entity={{id: 2}} case="variation-2">
        Variation 2
      </Flag>
      <Flag flag="treatment" entity={{id: 2}} case="off">
        Off
      </Flag>
    </>
  )

  expect(component.toJSON()).toMatchSnapshot()
})
