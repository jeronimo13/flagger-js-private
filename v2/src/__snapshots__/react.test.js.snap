// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flag accepts entity prop 1`] = `"Variation 1"`;

exports[`Flag can inherit flag prop from FlagSwitch 1`] = `"On!"`;

exports[`Flag can override flag prop from FlagSwitch 1`] = `"Default"`;

exports[`FlagSwitch accepts entity prop 1`] = `"Off!"`;

exports[`withFlag adds flags.<name>.eligible property 1`] = `
<div>
  true
</div>
`;

exports[`withFlag adds flags.<name>.enabled property 1`] = `
<div>
  true
</div>
`;

exports[`withFlag adds flags.<name>.payload property 1`] = `
<div>
  {"a":"b"}
</div>
`;

exports[`withFlag adds flags.<name>.treatment property 1`] = `
<div>
  on
</div>
`;

exports[`withFlag wrapped components accept entity prop 1`] = `
<div>
  false
</div>
`;

exports[`withFlag wrapped components handles invalid flag name 1`] = `
<div>
  false
</div>
`;
